src/tactic/sort.lean

```diff
@@ -0,0 +1,142 @@
+/-
+Copyright (c) 2022 Arthur Paulino, Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Arthur Paulino, Damiano Testa
+-/
+import data.polynomial.degree.definitions
+
+/-!  # A tactic for sorting sums  -/
+
+namespace tactic
+
+/--  Takes an `expr` and returns a list of its summands. -/
+meta def get_summands : expr → list expr
+| `(%%a + %%b) := get_summands a ++ get_summands b
+| a            := [a]
+
+section with_weights
+variables {N : Type*} [has_le N] [decidable_rel (has_le.le : N → N → Prop)]
+
+meta def compare_fn (wt : expr → option N) (eₗ eᵣ : expr) : bool :=
+match (wt eₗ, wt eᵣ) with
+| (some l, some r) := l ≤ r
+| (none, some _)   := tt
+| (some _, none)   := ff
+| _                := eₗ.to_string ≤ eᵣ.to_string -- this solution forces an unique ordering
+end
+
+/--  Given an expression `e` and a weight function `wt : expr → N` to a Type `N` with `<`,
+`sort_summands_with_weight a wt` returns the list of summands appearing in `a`, sorted using the
+order `<`. -/
+meta def sort_summands_with_weight (e : expr) (wt : expr → option N) : list expr :=
```

Comment:
I am a little bit uneasy about hard-coding that the order is on `option N`.  I'm happy to use a similar ordering in our applications, but `option N` with `compare_fn` is a *special* kind of type with `≤`, so is included as a special case in the def with a general `N`.  I would be happier if we redefined the order that we use on `monomials`, rather than crippling the general ordering possibilities.